light-client/src/peer_list.rs

```diff
@@ -1,156 +1,206 @@
 use crate::{
     bail,
     errors::{Error, ErrorKind},
-    supervisor::Instance,
     types::PeerId,
 };
 
 use contracts::{post, pre};
-use std::collections::{HashMap, HashSet};
+use std::collections::{BTreeSet, HashMap};
 
-/// A mapping from PeerIds to Light Client instances.
-/// Keeps track of which peer is deemed the primary peer.
+/// A generic container mapping `PeerId`s to some type `T`,
+/// which keeps track of the primary peer, witnesses, full nodes,
+/// and faulty nodes. Provides lifecycle methods to swap the primary,
+/// mark witnesses as faulty, and maintains an `invariant` for
+/// correctness.
 #[derive(Debug)]
-pub struct PeerList {
-    instances: HashMap<PeerId, Instance>,
+pub struct PeerList<T> {
+    values: HashMap<PeerId, T>,
     primary: PeerId,
-    witnesses: HashSet<PeerId>,
-    full_nodes: HashSet<PeerId>,
-    faulty_nodes: HashSet<PeerId>,
+    witnesses: BTreeSet<PeerId>,
```

Comment:
Maybe I missed some context in a commit, but what's the motivation to move to a `BTreeSet`? Asking purely out of curioisity.

Reply:
This is to provide a stable ordering on the peer ids, to make the whole process (more) deterministic.

Good question by the way, I failed to mention that in the commit.